### PR TITLE
[serve] Clear `replica_updated_event` in streaming router to avoid blocking proxy event loop

### LINKED ISSUE #36460: [serve][streaming] No replicas available blocks the HTTP proxy event loop
The loop waiting for a replica is not properly yielding. Repro:
```
# test.py
import asyncio

from ray import serve

from fastapi import FastAPI, WebSocket, WebSocketDisconnect
from starlette.responses import StreamingResponse
from starlette.requests import Request

app = FastAPI()

@serve.deployment()
@serve.ingress(app)
class A:
    def __init__(self):
        print("SLEEPING")
        import time;time.sleep(10)
        print("DONE SLEEPING")

    def __call__(self, *args):
        return "hi"

a = A.bind()
```

In one terminal:
```
RAY_SERVE_ENABLE_EXPERIMENTAL_STREAMING=1 serve run test:a
```

In second terminal:
```
curl -v -X GET http://localhost:8000/
curl -v -X GET http://localhost:8000/-/routes
```

The `/-/routes` call hangs until the replica finishes initializing.

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

`replica_updated_event` was not being cleared. It seems that even though we were `await`ing the event, it was not yielding the loop because the event was always set. This cause the `while replica is not None` loop to busy-spin, blocking the HTTP proxy event loop.

Closes https://github.com/ray-project/ray/issues/36460

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/http_proxy.py [hunk 1]
```diff
@@ -178,6 +178,12 @@ def __init__(self, controller_name: str):
         self.self_actor_handle = ray.get_runtime_context().current_actor
         self.asgi_receive_queues: Dict[str, ASGIMessageQueue] = dict()
 
+        if RAY_SERVE_ENABLE_EXPERIMENTAL_STREAMING:
```
Comment (GeneDer) [inline python/ray/serve/_private/http_proxy.py:181]: Nice logging!

**Review by GeneDer — APPROVED** [review]
LGTM!

File: python/ray/serve/_private/router.py [hunk 2]
```diff
@@ -121,6 +127,15 @@ async def assign_replica(
         )
 
     def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
+        new_replica_ids = {r.replica_tag for r in running_replicas}
+        if new_replica_ids != self._replica_id_set:
```
Comment (sihanwang41) [inline python/ray/serve/_private/router.py:131]: why do we get the replica update without change? I am not sure why we have to check here.

Comment (edoakes) [reply python/ray/serve/_private/router.py:131]: this is purely for logging

Comment (edoakes) [reply python/ray/serve/_private/router.py:131]: we might get spurious updates or the update might only be a config change

**Pull request merged**